File: frontend/src/utils/displayLabels.ts
type ProductLike = {
  product_id?: unknown;
  product_name?: unknown;
  evidence?: Record<string, unknown>;
};

export function formatProductLabel(productName?: unknown, productId?: unknown) {
  const name = String(productName ?? "").trim();
  const id = String(productId ?? "").trim();

  if (!id || id.toUpperCase() === "ALL") {
    return name || "All Products";
  }

  if (!name) return id;
  if (name.toLowerCase().includes(`(${id.toLowerCase()})`)) return name;

  return `${name} (${id})`;
}

export function productLabelFromFinding(finding?: ProductLike) {
  if (!finding) return "All Products";

  return formatProductLabel(
    finding.product_name ?? finding.evidence?.product_name,
    finding.product_id,
  );
}

export function replaceProductIds(
  text: string,
  products: Array<ProductLike>,
) {
  return products.reduce((result, product) => {
    const id = String(product.product_id ?? "").trim();
    const label = productLabelFromFinding(product);
    if (!id || id.toUpperCase() === "ALL" || label === id) return result;

    return result.replace(new RegExp(`\\b${id.replace(/[.*+?^${}()|[\\]\\]/g, "\\$&")}\\b`, "gi"),
      (match, offset, source) => source[offset - 1] === "(" ? match : label,
    );
  }, text);
}

export function formatRepresentativeLabel(
  representativeName?: unknown,
  representativeId?: unknown,
) {
  const name = String(representativeName ?? "").trim();
  const id = String(representativeId ?? "").trim();

  if (!id) return name || "Representative";
  if (!name) return `Representative (${id})`;
  if (name.toLowerCase().includes(`(${id.toLowerCase()})`)) return name;

  return `${name} (${id})`;
}

export function replaceRepresentativeId(
  text: string,
  representativeName?: unknown,
  representativeId?: unknown,
) {
  const id = String(representativeId ?? "").trim();
  if (!id) return text;

  const label = formatRepresentativeLabel(representativeName, id);
  return text.replace(new RegExp(`\\b${id.replace(/[.*+?^${}()|[\\]\\]/g, "\\$&")}\\b`, "gi"),
    (match, offset, source) => source[offset - 1] === "(" ? match : label,
  );
}

File: backend/app/utils/display_labels.py
from __future__ import annotations

import re
from typing import Any


def product_label(product_name: Any, product_id: Any) -> str:
    name = str(product_name or "").strip()
    identifier = str(product_id or "").strip()
    if not identifier or identifier.upper() == "ALL":
        return name or "All Products"
    if not name:
        return f"Product ({identifier})"
    if f"({identifier.lower()})" in name.lower():
        return name
    return f"{name} ({identifier})"


def representative_label(representative_name: Any, representative_id: Any) -> str:
    name = str(representative_name or "").strip()
    identifier = str(representative_id or "").strip()
    if not identifier:
        return name or "Representative"
    if not name:
        return f"Representative ({identifier})"
    if f"({identifier.lower()})" in name.lower():
        return name
    return f"{name} ({identifier})"


def investigation_text(value: Any, state: dict[str, Any]) -> str:
    text_value = str(value)
    findings = state.get("findings", []) or []

    for finding in findings:
        product_id = str(finding.get("product_id") or "").strip()
        if not product_id or product_id.upper() == "ALL":
            continue
        product_name = finding.get("product_name") or (finding.get("evidence") or {}).get(
            "product_name"
        )
        label = product_label(product_name, product_id)
        text_value = re.sub(
            rf"(?<!\()\b{re.escape(product_id)}\b",
            lambda _match: label,
            text_value,
            flags=re.IGNORECASE,
        )

    representative_id = str(state.get("representative_id") or "").strip()
    representative_name = state.get("representative_name")
    if not representative_name:
        representative = next(
            (
                item
                for item in state.get("representatives", []) or []
                if item.get("representative_id") == representative_id
            ),
            None,
        )
        if representative:
            representative_name = " ".join(
                part
                for part in (
                    str(representative.get("first_name") or "").strip(),
                    str(representative.get("last_name") or "").strip(),
                )
                if part
            )

    if representative_id:
        label = representative_label(representative_name, representative_id)
        text_value = re.sub(
            rf"(?<!\()\b{re.escape(representative_id)}\b",
            lambda _match: label,
            text_value,
            flags=re.IGNORECASE,
        )

    return text_value

